File: 101-250/105-construct-binary-tree-from-preorder-and-inorder-traversal.cpp
class Solution {
    unordered_map<int, int> hash;
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        for (int i = 0; i < inorder.size(); ++i) {
            hash[inorder[i]] = i;
        }
        return helper(preorder, 0, inorder.size()-1, 0);
    }

    TreeNode* helper(vector<int>& preorder, int inStart, int inEnd, int preStart) {
        if (inStart > inEnd) return nullptr;
        int rootVal = preorder[preStart];
        int inMid = hash[rootVal];
        int leftLen = inMid - inStart;
        TreeNode *root = new TreeNode(rootVal);
        root->left = helper(preorder, inStart, inMid - 1, preStart + 1);
        root->right = helper(preorder, inMid + 1, inEnd, preStart + 1 + leftLen);
        return root;
    }
};

File: python version/105-construct-binary-tree-from-preorder-and-inorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        
        def helper(pre_start, in_start, in_end):
            if in_start > in_end: return None
            val = preorder[pre_start]
            node = TreeNode(val)
            in_pos = inorder.index(val)
            len_left = in_pos - in_start
            node.left = helper(pre_start+1, in_start, in_pos-1)
            node.right = helper(pre_start+1+len_left, in_pos+1, in_end)
            return node
        
        return helper(0, 0, len(inorder)-1)